bundle/src/main/java/com/adobe/acs/commons/remoteassets/impl/RemoteAssetDecorator.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * #%L
+ * ACS AEM Commons Bundle
+ * %%
+ * Copyright (C) 2018 Adobe
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package com.adobe.acs.commons.remoteassets.impl;
+
+import com.adobe.acs.commons.remoteassets.RemoteAssetsBinarySync;
+import com.adobe.acs.commons.remoteassets.RemoteAssetsConfig;
+import org.apache.felix.scr.annotations.Component;
+import org.apache.felix.scr.annotations.ConfigurationPolicy;
+import org.apache.felix.scr.annotations.Reference;
+import org.apache.felix.scr.annotations.Service;
+import org.apache.jackrabbit.api.security.user.User;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ResourceDecorator;
+import org.apache.sling.api.resource.ValueMap;
+import org.apache.sling.jcr.base.util.AccessControlUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.jcr.RepositoryException;
+import javax.jcr.Session;
+import javax.servlet.http.HttpServletRequest;
+import java.util.Calendar;
+import java.util.Set;
+import java.util.concurrent.ConcurrentSkipListSet;
+
+/**
+ * ResourceDecorator that instruments remote assets to sync binaries as needed.
+ * This "decorator" is used to detect the first time a "remote" asset is
+ * referenced by the system and sync that asset from the remote server to
+ * make it now a "true" asset.
+ */
+@Component(
+        label = "ACS AEM Commons - Remote Assets - Asset Resource Decorator",
+        description = "Captures a request for a remote asset so that the binary can be sync'd to the current server making it a true local asset",
+        policy = ConfigurationPolicy.REQUIRE
+)
+@Service
+public class RemoteAssetDecorator implements ResourceDecorator {
```

Comment:
Are there situations where this doesnt work? IIRC the Assets APIs are generally lose w/ always using Sling Resource APIs and often drop straight to the JCR Node APIs?

Reply:
It's a good question - LMK a scenario you want to test.  I have yet to run into a valid scenario where I would expect it to work and it didnt.  In the case of Asset APIs returning JSON it would probably be ideal that it *doesn't* work b/c all of the metadata and nodes are there for a correct JSON API return, and then if someone actually accesses the asset to view it then at that time it would sync.